src/node_quic_buffer.h

```diff
@@ -0,0 +1,255 @@
+  // of uv_buf_t instances. The destination must be
+  // an empty QuicBuffer.
+  inline void TransferAndReset(QuicBuffer* dest) {
+    CHECK(!dest->root_);
+    CHECK_EQ(0, length_);
+    CHECK_EQ(0, size_);
+    dest->head_ = head_;
+    dest->tail_ = tail_;
+    dest->size_ = size_;
+    dest->length_ = length_;
+    head_ = nullptr;
+    tail_ = nullptr;
+    size_ = nullptr;
+    length_ = nullptr;
+    dest->root_.reset(root_.release());
+  }
+
+  // Push one or more uv_buf_t instances into the buffer.
+  // the done_cb callback will be invoked when the last
+  // uv_buf_t in the bufs array is consumed and popped out
+  // of the internal deque. The user_data is passed in to
+  // the done_cb. The keep_alive allows a reference to a
+  // JS object to be kept around until the final uv_buf_t
+  // is consumed.
+  inline void Push(
+      uv_buf_t* bufs,
+      size_t nbufs,
+      done_cb done = default_quic_buffer_chunk_done,
+      void* user_data = nullptr,
+      v8::Local<v8::Object> keep_alive_ = v8::Local<v8::Object>()) {
+    if (nbufs == 0)
+      return done(0, 0, user_data);
+    size_t n = 0;
+    while (nbufs > 1) {
+      Push(bufs[n]);
+      length_ += bufs[n].len;
+      n++;
+      nbufs--;
+    }
+    length_ += bufs[n].len;
+    Push(bufs[n], done, user_data, keep_alive_);
+  }
+
+  // Consume the given number of bytes within the buffer. If amount is
+  // negative, all buffered bytes are consumed.
+  inline void Consume(ssize_t amount = -1) { Consume(0, amount); }
+
+  // Cancels the remaining bytes within the buffer
+  inline void Cancel() { Consume(UV_ECANCELED, -1); }
+
+  // The total buffered bytes
+  inline uint64_t Length() {
+    return length_;
+  }
+
+  // The total number of buffers
+  inline size_t Size() {
+    return size_;
+  }
+
+  // Drain the remaining buffers into the given vector
+  inline size_t DrainInto(std::vector<uv_buf_t*>* list) {
+    uv_buf_t* head = Head();
+    size_t len = 0;
+    while (head != nullptr) {
+      list->push_back(head);
+      SeekHead();
+      len++;
+    }
+    return len;
+  }
+
+  // Returns the current read head or an empty buffer if
+  // we're empty
+  inline uv_buf_t* Head() {
+    if (!head_)
+      return nullptr;
+    return &head_->buf;
+  }
+
+  // Moves the current read head forward the given
+  // number of buffers. If amount is greater than
+  // the number of buffers remaining, move to the
+  // end, and return the actual number advanced.
+  inline size_t SeekHead(size_t amount = 1) {
+    size_t n = 0;
+    while (amount > 0 && head_ != nullptr) {
+      head_ = head_->next.get();
+      n++;
+      amount--;
+    }
+    return n;
+  }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackFieldWithSize("length", length_);
+  }
+  SET_MEMORY_INFO_NAME(QuicBuffer);
+  SET_SELF_SIZE(QuicBuffer);
+
+ private:
+  inline void Push(quic_buffer_chunk* chunk) {
+    size_++;
+    if (!tail_) {
+      root_.reset(std::move(chunk));
```

Comment:
`std::move()` doesn’t do anything for pointers … is this intentional here?

Reply:
ah right, artifact of a prior change really that I hadn't changed back yet.